src/parser/mod.rs

```diff
@@ -471,9 +471,47 @@ fn term<'a, 'b>(input: Input<'a, 'b>) -> AstResult<'a, 'b> {
 parser!(function_call loc =
     function <- member_access;
     args <- many1(function_argument);
-    Ast::function_call(function, args, loc)
+    desugar_explicit_currying(function, args, loc)
 );
 
+/// Turns (foo _  _ 2) into (\ _$1 _$2 . (foo _$1 _$2 2))
+fn desugar_explicit_currying<'a>(function: Ast<'a>, args: Vec<Ast<'a>>, loc: Location<'a>) -> Ast<'a> {
+    if matches_not_typeconstructor(&function) && args.iter().any(matches_underscore) {
+        return curried_function_call(function, args, loc)
+    }
+
+    Ast::function_call(function, args, loc)
+}
+
+fn curried_function_call<'a>(function: Ast<'a>, args: Vec<Ast<'a>>, loc: Location<'a>) -> Ast<'a> {
+    let mut curried_args = vec![];
+    let mut curried_arg_count = 0;
+    let args: Vec<Ast<'a>> = args
+        .into_iter()
+        .map(|arg| {
+            if matches_underscore(&arg) {
+                curried_arg_count += 1;
+                let curried_arg = format!("_${}", curried_arg_count);
+                curried_args.push(Ast::variable(curried_arg.clone(), loc));
+                Ast::variable(curried_arg, loc)
+            } else {
+                arg
+            }
+        })
+        .collect();
+
+    let function_call = Ast::function_call(function, args, loc);
+    Ast::lambda(curried_args, None, function_call, loc)
+}
+
+fn matches_underscore(arg: &Ast) -> bool {
+    matches!(arg, Ast::Variable(ast::Variable{ kind: ast::VariableKind::Identifier(x), ..}) if x == "_")
+}
+
+fn matches_not_typeconstructor(function: &Ast) -> bool {
```

Comment:
Is this and the other `matches_underscore` function implemented somewhere else or is this alright? We could put it in an impl block by the enum definitions, but it might be a little overkill just for one place :)

Reply:
I think there's precedence for being in an impl block though (there are some existing functions there like `is_pair_constructor` I believe). Since these are only used for parsing I'm fine with them either staying here or being in an impl.

I don't think we need a `matches_not_typeconstructor` check though. Ideally we should be able to curry type constructors too. Something like the following should work:
```hs
type Person = name: string, job: string

names = ["alice", "bob", ...]
people = map names (Person _ "programmer")
```
Conceptually, type constructors are just functions so there's no need to special case them here. For example the `Person` type constructor has type `string string -> Person`

Reply:
I didn't have the check at first, but the desugaring logic triggered on match arms for match statements and resulted in
```
   | examples/typechecking/completeness_checking.an: 7,4        error: Type mismatch between (Maybe a) and (a -> (Maybe a))
   | | (_, Some _) -> ()
```
What should we do about this situation?

Reply:
Oh, looks like we need to not desugar if we're in a pattern match. This could complicate the parser some. Changing `irrefutable_pattern` would be easy but refutable patterns just re-use the `expr` parser which makes it harder. If you find a good way of including it then we can add that in but otherwise I'm fine with leaving this as a TODO for later.